src/MongoDB.Driver.Core/Core/Connections/TcpStreamFactory.cs

```
@@ -259,9 +259,11 @@ private Socket CreateSocket(EndPoint endPoint)
 
             var socket = new Socket(addressFamily, SocketType.Stream, ProtocolType.Tcp);
 
-            // not all platforms support IOControl
-            try
+            if (OperatingSystemHelper.CurrentOperatingSystem == OperatingSystemPlatform.Windows)
             {
+                // Reviewing the .NET source, Socket.IOControl for IOControlCode.KeepAlivesValue will
```

can we rely on undocumented behavior? As I understand this can be changed at further versions. Can you please explain what's the issue with the previous workflow? As I understand, both behaviors will be the same (for old and new versions) eventually

---

I agree with you that I don't relying on undocumented behaviour and that it could change in a future version. That said, this portion of the PAL hasn't changed in 5 years.

https://github.com/dotnet/runtime/blame/main/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs#L1346

The reason why is twofold:

First exceptions are a lot slower than if statements. A quick test shows that you can thrown ~48 exceptions per millisecond compared to evaluating 20K+ branches per millisecond. Given we are only catching a single exception per created connection and connections are being pooled, this is less of a concern than if we were in a tight loop. Still why add the extra delay when we know `Socket.IOControl` is going to fail on non-Windows systems?

The second reason is that some monitoring frameworks will report on exception behaviour including caught exceptions. Thus avoiding the expected `PlatformNotSupportedExceptions` on non-Windows platforms cleans up the monitoring stats, which I understand is the original ask in CSHARP-4287.

The overall behaviour before and after this proposed change should be identical, but connection creation will be slightly faster on non-Windows systems with less monitoring noise after the change.

Happy to discuss further if you still feel this is a bad idea.

---

I consider the previous logic after exception as fallback, which means using throwing exception as workflow is one of expected ways to do it. 

So my concerns:
1. Now we have different workflow. Previously any `PlatformNotSupportedException` was suppressed. We also additionally had a fallback mechanism that also always suppressed  `PlatformNotSupportedException`. Now windows doesn't suppress any exceptions.
2. There is a number of discussions like https://github.com/dotnet/runtime/issues/25555#issuecomment-375087989 or even approved PR like https://github.com/dotnet/corefx/pull/29963. For example, it looks like SetSocletOptions can be set for windows as well: https://github.com/dotnet/corefx/pull/29963/files#diff-d2470117f55a00b4e04f47d3411800d9c4b49df7619e21a0e1a8b48b6d93a4bdR52 
3. Can we rely say that `IOControl` always works only on windows? What about let's say mono on linux? Or some similar weird or legacy setup? What about old TFs?

---

Point #1...
The previous logic:
- On Windows, try `IOControl` - which would always succeed as all versions of Windows supports this. .NET Framework, .NET Core, and Mono all call `ioctl` Win32 API.
- On non-Windows, try IOControl, throw, try SetSocketOption for KeepAlive, which succeeds on all platforms (according to Stephen Toub https://github.com/dotnet/runtime/issues/25555#issuecomment-375087989). Even if it fails, we catch the `PlatformNotSupportedException` and ignore. 

The new logic:
- On Windows, set via `IOControl`. We don't fall back to `SetSocketOption`, but we don't need to because `IOControl` will always succeed on Windows.
- On non-Windows, skip `IOControl` (because we know it will fail) and use `SetSocketOption`. Although it should work on all platforms (according to Stephen Toub above), we catch `PlatformNotSupportedException` as before and ignore.

In summary the new logic is equivalent to the old for all practical purposes.

Point #2...
`SetSocketOption` can be used on Windows but all it does is enable keep alives, whereas `IOControl` enables keep alives and sets both the keepalives time and internal. So if `IOControl` works - which it will on Windows - it does everything that `SetSocketOption` does and more.

The more interesting point is other .NET impls like Mono. Mono partially implements `IOControl` on non-Windows operating systems:

```
/* See Socket.IOControl, WSAIoctl documentation in MSDN. The common options between UNIX
 * and Winsock are FIONREAD, FIONBIO and SIOCATMARK. Anything else will depend on the system
 * except SIO_KEEPALIVE_VALS which is properly handled on both windows and linux. */
```

See https://github.com/mono/mono/blob/main/mcs/class/System/System.Net.Sockets/Socket.cs#L2763-L2765.

We don't officially support Mono, nor do we test on it. So we shouldn't be basing our decision of whether to call `Socket.IOControl` due to partial support in Mono.

Point #3...
Mono partially implements `IOControl`. Some future .NET 7, 8, 9, ???, ... might support it on non-Windows operating systems. But historically it hasn't supported it in the last 5 years nor have any changes been made in this code in the last 5 years. If a future version does support it, we can always remove the platform check.

As for other TFs... .NET Framework only runs on Windows. So that's a moot point. We support back to .NET Core 2.1, which doesn't support `IOControl` on non-Windows. .NET 6 (latest) doesn't support `IOControl` on non-Windows. Mono partially supports it, but see point 2 above.

Does that sufficiently address your questions/concerns? Or are you still hesitant about this change?